R/dummy_data.R

```diff
@@ -1256,6 +1345,330 @@ dummy_er <- function(nid = 1000, n_hospitals = 10, time_period = c(2015, 2023),
+  if (!is.null(seed)) {
+    set.seed(seed)
+  }
+
+  ####### checks for valid inputs #######
+  if (!is.null(cohort)) { # if `cohort` is provided
+    check_input(cohort,
+      c("data.frame", "data.table"),
+      colnames = c("genc_id", "hospital_num"),
+      coltypes = c("integer", "integer")
+    )
+  } else { # when `cohort` is not provided
+    check_input(list(nid, n_hospitals), "integer")
+
+    if (nid < n_hospitals) {
+      stop("Number of encounters must be greater than or equal to the number of hospitals")
+    }
+  }
+
+  if (!is.null(cohort)) {
+    # if `cohort` is provided, use its `genc_id` and `hospital_num`
+    cohort <- as.data.table(cohort)
+    df1 <- generate_id_hospital(cohort = cohort, include_prop = 1, avg_repeats = 1, seed = seed)
+
+    nid <- length(unique(df1$genc_id))
+    n_hospitals <- length(unique(df1$hospital_num))
+
+    # only include the genc_id and hospital_num columns from `cohort`
+    df1 <- df1[, c("genc_id", "hospital_num")]
+  } else {
+    # no repeating genc_id in physicians data table
+    df1 <- generate_id_hospital(nid, n_hospitals, avg_repeats = 1, seed = seed)
+  }
+
+  # Function that samples a value from a given vector with replacement
+  sample_from_col <- function(dat) {
+    return(sample(dat, 1, replace = TRUE))
+  }
+
+  # set the NA proportions per hospital per variable
+  # sampling creates hospital-level variation
+  # each row is a different hopsital
+  hosp_na_prop <- df1 %>%
+    group_by(
+      hospital_num
+    ) %>%
+    reframe(
+      admitting_phy_gim_NA = runif(1, 0, 1),
+      discharging_phy_gim_NA = runif(1, 0, 1),
+      adm_phy_cpso_mapped_NA = rbeta(1, 0.2, 1.1),
+      dis_phy_cpso_mapped_NA = rbeta(1, 0.25, 2.7),
+      mrp_cpso_mapped_NA = rbeta(1, 0.4, 11.2),
+      admitting_phy_gim_NA = runif(1, 0, 1),
+      admitting_phy_gim_Y = runif(1, 0, 1),
+      discharging_phy_gim_NA = runif(1, 0, 1),
+      discharging_phy_gim_Y = runif(1, 0, 1),
+    ) %>%
+    data.table()
+
+  # edit `adm_phy_cpso_mapped` and `dis_phy_cpso_mapped`
+  # hospitals with highest adm also have highest dis NA proportion
+  # artificially add some counts of proportion 1.0 NA
+  na_order_adm <- order(hosp_na_prop[["adm_phy_cpso_mapped_NA"]], decreasing = TRUE)
+  hosp_na_prop[["dis_phy_cpso_mapped_NA"]][na_order_adm[1:round(0.15 * n_hospitals)]] <- jitter(
+    hosp_na_prop[["dis_phy_cpso_mapped_NA"]][na_order_adm[1:round(0.15 * n_hospitals)]],
+    factor = 0.1
+  )
+
+  hosp_na_prop[["adm_phy_cpso_mapped_NA"]][na_order_adm[1:round(0.07 * n_hospitals)]] <- 1
+  hosp_na_prop[["dis_phy_cpso_mapped_NA"]][na_order_adm[1:round(0.07 * n_hospitals)]] <- 1
+
+  # sample all physician numbers
+  # each hospital has about 280 physicians on average
+  # multiply this average by n_hospitals to get the total set of physicians across all hospitals
+  sample_cpso <- round(runif(280 * n_hospitals, min = 1e4, max = 3e5))
```

Comment:
since the randomly generated numbers can actually generate values collide with real cspo numbers - this may not be very ideal. Perhaps we should be explicit that these are fake cpsos. Could we add a prefix "SYN_" to signal that these are synthetic IDs as below:
```suggestion
  sample_cpso <- paste0("SYN_", round(runif(280 * n_hospitals, min = 1e4, max = 3e5)))
```